File: EmployeeShifts.py
import random
from collections import defaultdict, Counter
from typing import List, Dict, Tuple

DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
SHIFTS = ["morning", "afternoon", "evening"]

class Employee:
    def __init__(self, name: str):
        self.name = name
        # preferences[day_index] = list of shifts in priority order, e.g. ["morning","evening"]
        self.preferences: Dict[int, List[str]] = {d: [] for d in range(7)}
        self.days_assigned = 0
        # assigned_shifts[day_index] = shift_name or None
        self.assigned_shifts: Dict[int, str] = {}

    def __repr__(self):
        return f"Employee({self.name}, assigned_days={self.days_assigned})"

def ask_user_input() -> Tuple[List[Employee], int]:
    """
    Interactive input helper. User can choose to enter employees manually or use sample data.
    Returns (employees_list, max_per_shift)
    """
    print("Employee Shift Scheduler — Input")
    use_sample = input("Use sample data? (y/n) [n]: ").strip().lower() or "n"
    employees = []
    if use_sample == "y":
        # Create sample employees with random preferences
        sample_names = ["Alice", "Bob", "Charlie", "Deepa", "Ethan", "Farah", "Gopal", "Hina", "Irfan"]
        employees = [Employee(n) for n in sample_names]
        for e in employees:
            for d in range(7):
                # create a random ranking of SHIFTS
                prefs = SHIFTS[:] 
                random.shuffle(prefs)
                # keep full ranking (bonus)
                e.preferences[d] = prefs
        print(f"Created {len(employees)} sample employees.")
    else:
        n = int(input("How many employees? Enter integer: ").strip())
        for i in range(n):
            name = input(f"Name of employee #{i+1}: ").strip() or f"Emp{i+1}"
            emp = Employee(name)
            print(f"Now enter preferences for {name}. For each day provide ranked shifts separated by '>' (e.g. morning>evening) or comma.")
            print("Use any subset of ['morning','afternoon','evening']. If empty, we treat as 'no preference'.")
            for d_idx, day in enumerate(DAYS):
                raw = input(f"  {day} preferences: ").strip()
                if not raw:
                    emp.preferences[d_idx] = []
                else:
                    # normalize input: separators '>' or ',' or whitespace
                    parts = [p.strip().lower() for p in raw.replace(">", ",").split(",") if p.strip()]
                    # filter invalid names and keep order
                    emp.preferences[d_idx] = [p for p in parts if p in SHIFTS]
            employees.append(emp)
    # max per shift
    max_per_shift_str = input("Enter maximum employees allowed per shift (per day). Default 5: ").strip()
    max_per_shift = int(max_per_shift_str) if max_per_shift_str.isdigit() else 5
    return employees, max_per_shift

def initialize_schedule():
    # schedule[day_idx][shift] = list of employee names assigned
    schedule = [ {s: [] for s in SHIFTS} for _ in range(7) ]
    return schedule

def can_assign(emp: Employee, day: int) -> bool:
    """Check if employee can be assigned on given day (not already assigned that day and under 5 days)."""
    if day in emp.assigned_shifts:
        return False
    if emp.days_assigned >= 5:
        return False
    return True

def assign_employee_to_shift(emp: Employee, day: int, shift: str, schedule: List[Dict[str,List[str]]]):
    schedule[day][shift].append(emp.name)
    emp.assigned_shifts[day] = shift
    emp.days_assigned += 1

def schedule_with_preferences(employees: List[Employee], max_per_shift: int):
    schedule = initialize_schedule()

    # Randomize employee order per day for fairness
    emp_order = employees[:]
    random.shuffle(emp_order)

    # First pass: try to assign employees to their top available preference for each day
    for day in range(7):
        random.shuffle(emp_order)
        for emp in emp_order:
            if not can_assign(emp, day):
                continue
            prefs = emp.preferences.get(day, []) or []
            assigned = False
            # try each preference in ranked order
            for shift in prefs:
                if len(schedule[day][shift]) < max_per_shift:
                    assign_employee_to_shift(emp, day, shift, schedule)
                    assigned = True
                    break
            # if no preferences or preferred full, we'll try in conflict resolution later (or now try other shifts same day)
            if not assigned:
                # attempt other shifts on same day (non-preferred) as early conflict resolution
                for shift in SHIFTS:
                    if shift in prefs:
                        continue
                    if len(schedule[day][shift]) < max_per_shift:
                        assign_employee_to_shift(emp, day, shift, schedule)
                        assigned = True
                        break
            # if still not assigned, leave unassigned for now (maybe next-day resolution)
    # After first pass, ensure min staffing per shift per day
    shortages = ensure_minimum_staffing(schedule, employees, min_required=2, max_per_shift=max_per_shift)

    # For employees still with free capacity (under 5 days), try to fill any remaining unassigned employees who are not assigned any shift on certain days
    # Also try to resolve any unfilled shift shortages by looking ahead next days if possible.
    if shortages:
        resolve_shortages_via_next_day(schedule, employees, shortages, max_per_shift)

    # Final check: try to assign unassigned employees (who still can work days) to any shift minimally to maximize coverage
    fill_remaining_with_available(schedule, employees, max_per_shift)

    return schedule

def ensure_minimum_staffing(schedule: List[Dict[str,List[str]]], employees: List[Employee], min_required: int, max_per_shift: int):
    """
    For each day & shift ensure at least min_required employees assigned.
    If not enough prefer that shift, randomly pick eligible employees who:
    - are not assigned that day
    - have days_assigned < 5
    Preference is given to those who haven't been assigned much.
    Returns a list of shortages as tuples (day, shift, missing_count).
    """
    shortages = []
    # Build a name->employee map
    emp_map = {e.name: e for e in employees}
    for day in range(7):
        for shift in SHIFTS:
            need = min_required - len(schedule[day][shift])
            if need > 0:
                # choose eligible employees
                eligible = [e for e in employees if can_assign(e, day)]
                # sort eligible by fewest days_assigned to spread load
                eligible.sort(key=lambda x: x.days_assigned)
                chosen = []
                for e in eligible:
                    if len(chosen) >= need:
                        break
                    # ensure we don't exceed max_per_shift
                    if len(schedule[day][shift]) + len(chosen) < max_per_shift:
                        chosen.append(e)
                for e in chosen:
                    assign_employee_to_shift(e, day, shift, schedule)
                short_remaining = need - len(chosen)
                if short_remaining > 0:
                    shortages.append((day, shift, short_remaining))
    return shortages

def resolve_shortages_via_next_day(schedule: List[Dict[str,List[str]]], employees: List[Employee], shortages: List[Tuple[int,str,int]], max_per_shift:int):
    """
    Try to resolve shortages by moving employees from other days (who have flexible preferences)
    or assigning employees to adjacent days/shifts if they have capacity.
    This is a heuristic: we try to find employees who are free on the shortage day and have capacity.
    """
    if not shortages:
        return []
    emp_map = {e.name: e for e in employees}
    updated_shortages = []
    for day, shift, need in shortages:
        chosen = []
        eligible = [e for e in employees if can_assign(e, day)]
        eligible.sort(key=lambda x: x.days_assigned)
        for e in eligible:
            if len(chosen) >= need:
                break
            if len(schedule[day][shift]) + len(chosen) < max_per_shift:
                assign_employee_to_shift(e, day, shift, schedule)
                chosen.append(e)
        still = need - len(chosen)
        if still > 0:
            # try next days: assign someone to next day but we need staffing for this specific day,
            # so next-day assignment doesn't fix today's shortage. We'll mark as unresolved.
            updated_shortages.append((day, shift, still))
    return updated_shortages

def fill_remaining_with_available(schedule: List[Dict[str,List[str]]], employees: List[Employee], max_per_shift: int):
    """
    Final pass: assign any employees who still have capacity to days where they were unassigned and shift has space.
    Tries to respect their preferences first.
    """
    emp_map = {e.name: e for e in employees}
    # For each day, for each employee unassigned that day and with capacity, try to assign to a space
    for day in range(7):
        for emp in employees:
            if not can_assign(emp, day):
                continue
            prefs = emp.preferences[day]
            assigned = False
            # try preferences first
            for shift in prefs:
                if len(schedule[day][shift]) < max_per_shift:
                    assign_employee_to_shift(emp, day, shift, schedule)
                    assigned = True
                    break
            if assigned:
                continue
            # try any shift
            for shift in SHIFTS:
                if len(schedule[day][shift]) < max_per_shift:
                    assign_employee_to_shift(emp, day, shift, schedule)
                    break

def pretty_print_schedule(schedule: List[Dict[str,List[str]]]):
    print("\nFinal Schedule for the Week:\n")
    for d_idx, day in enumerate(DAYS):
        print(f"{day}:")
        for shift in SHIFTS:
            names = schedule[d_idx][shift]
            print(f"  {shift.title():9}: {', '.join(names) if names else '(none)'}")
        print("-" * 40)

def summary_stats(schedule: List[Dict[str,List[str]]], employees: List[Employee]):
    # Count total assignments per employee and report anyone who exceeded constraints (shouldn't happen)
    counts = Counter()
    for d in range(7):
        for s in SHIFTS:
            for name in schedule[d][s]:
                counts[name] += 1
    print("\nEmployee assignment summary (days assigned):")
    for e in employees:
        print(f" - {e.name:12}: {e.days_assigned} day(s)")
    # Check for issues
    issues = []
    for e in employees:
        if e.days_assigned > 5:
            issues.append(f"{e.name} assigned > 5 days ({e.days_assigned})")
    # Check shifts for min staffing
    for d in range(7):
        for s in SHIFTS:
            if len(schedule[d][s]) < 2:
                issues.append(f"{DAYS[d]} {s} has fewer than 2 staff ({len(schedule[d][s])})")
    if issues:
        print("\nWarnings / Issues detected:")
        for it in issues:
            print(" -", it)
    else:
        print("\nNo issues detected. All constraints satisfied (or warnings handled).")

def main():
    employees, max_per_shift = ask_user_input()
    if not employees:
        print("No employees provided. Exiting.")
        return
    print("\nScheduling... (this may reshuffle employees for fairness)")
    schedule = schedule_with_preferences(employees, max_per_shift)
    pretty_print_schedule(schedule)
    summary_stats(schedule, employees)
    # Optionally export to file
    save = input("\nWould you like to save this schedule to 'schedule_week.txt'? (y/n) [n]: ").strip().lower() or "n"
    if save == "y":
        with open("schedule_week.txt", "w") as f:
            for d_idx, day in enumerate(DAYS):
                f.write(f"{day}:\n")
                for shift in SHIFTS:
                    names = schedule[d_idx][shift]
                    f.write(f"  {shift.title():9}: {', '.join(names) if names else '(none)'}\n")
                f.write("\n")
        print("Saved to schedule_week.txt")

if __name__ == "__main__":
    main()

File: EmployeeShifts.java
import java.util.*;
import java.util.stream.Collectors;

/**
 * Employee Shift Scheduling System
 * - Collects employee weekly preferences (primary + optional secondary)
 * - Schedules shifts for 7 days (MON..SUN) with Morning/Afternoon/Evening shifts
 * - Enforces constraints:
 *     - Max 1 shift per employee per day
 *     - Max 5 working days per employee per week
 *     - Minimum 2 employees per shift per day; fills randomly if needed
 * - Resolves conflicts when preferred shift is full by trying secondary or next day
 *
 * Assumptions:
 * - To create meaningful "full shift" conflicts, we use:
 *     MIN_PER_SHIFT = 2 (company requirement)
 *     MAX_PER_SHIFT = 3 (capacity per shift)  <-- change this if you want different behavior
 *
 * The program is interactive; enter 0 employees to run a built-in sample dataset.
 */
public class EmployeeScheduler {

    // Constants
    static final int MIN_PER_SHIFT = 2;
    static final int MAX_PER_SHIFT = 3; // capacity per shift (used to model "full" shift conflicts)
    static final int MAX_DAYS_PER_EMPLOYEE = 5;

    enum Day {
        MONDAY, TUESDAY, WEDNESDAY, THURSDAY, FRIDAY, SATURDAY, SUNDAY;

        public static Day[] week() { return values(); }
    }

    enum Shift {
        MORNING, AFTERNOON, EVENING, NONE; // NONE means unavailable / no preference
        public static Shift fromChar(char c) {
            c = Character.toUpperCase(c);
            switch (c) {
                case 'M': return MORNING;
                case 'A': return AFTERNOON;
                case 'E': return EVENING;
                case 'N': return NONE;
                default: return NONE;
            }
        }
    }

    static class Preference {
        Shift primary;
        Shift secondary; // may be NONE if not provided

        Preference(Shift primary, Shift secondary) {
            this.primary = primary;
            this.secondary = secondary;
        }
    }

    static class Employee {
        String name;
        Map<Day, Preference> preferences = new EnumMap<>(Day.class);
        Map<Day, Shift> assigned = new EnumMap<>(Day.class);
        int assignedDays = 0;

        Employee(String name) { this.name = name; 
            for (Day d : Day.week()) assigned.put(d, Shift.NONE);
        }

        boolean isAssignedOn(Day d) {
            return assigned.get(d) != Shift.NONE;
        }

        void assign(Day d, Shift s) {
            if (assigned.get(d) == Shift.NONE && s != Shift.NONE) {
                assigned.put(d, s);
                assignedDays++;
            } else if (assigned.get(d) != Shift.NONE && s == Shift.NONE) {
                assigned.put(d, Shift.NONE);
            } else {
                assigned.put(d, s); // overwrite (generally not used)
            }
        }

        boolean canWorkMoreDays() {
            return assignedDays < MAX_DAYS_PER_EMPLOYEE;
        }

        @Override
        public String toString() { return name; }
    }

    static class Schedule {
        // Map: Day -> Shift -> List<Employee>
        Map<Day, Map<Shift, List<Employee>>> schedule = new EnumMap<>(Day.class);

        Schedule() {
            for (Day d : Day.week()) {
                Map<Shift, List<Employee>> m = new EnumMap<>(Shift.class);
                for (Shift s : new Shift[]{Shift.MORNING, Shift.AFTERNOON, Shift.EVENING})
                    m.put(s, new ArrayList<>());
                schedule.put(d, m);
            }
        }

        List<Employee> get(Day d, Shift s) {
            return schedule.get(d).get(s);
        }

        void add(Day d, Shift s, Employee e) {
            schedule.get(d).get(s).add(e);
            e.assign(d, s);
        }

        int count(Day d, Shift s) {
            return schedule.get(d).get(s).size();
        }

        String prettyPrint() {
            StringBuilder sb = new StringBuilder();
            for (Day d : Day.week()) {
                sb.append(d.name()).append(":\n");
                for (Shift s : new Shift[]{Shift.MORNING, Shift.AFTERNOON, Shift.EVENING}) {
                    sb.append("  ").append(s.name()).append(": ");
                    List<Employee> list = get(d,s);
                    if (list.isEmpty()) sb.append("[none]");
                    else sb.append(list.stream().map(emp -> emp.name).collect(Collectors.joining(", ")));
                    sb.append("\n");
                }
                sb.append("\n");
            }
            return sb.toString();
        }
    }

    // Utility random
    static Random rand = new Random();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("Employee Shift Scheduling System");
        System.out.println("--------------------------------");
        System.out.print("Enter number of employees (0 to run sample dataset): ");
        int n = readIntSafe(sc);

        List<Employee> employees = new ArrayList<>();

        if (n <= 0) {
            employees = sampleEmployees();
            System.out.println("\nRunning sample dataset with employees:");
            employees.forEach(e -> System.out.println(" - " + e.name));
            System.out.println();
        } else {
            for (int i = 0; i < n; ++i) {
                System.out.print("\nEnter name of employee #" + (i+1) + ": ");
                String name = sc.nextLine().trim();
                if (name.isEmpty()) { System.out.println("Name can't be empty. Try again."); i--; continue; }
                Employee emp = new Employee(name);
                for (Day d : Day.week()) {
                    System.out.println("For " + emp.name + ", day: " + d.name());
                    System.out.print("  Primary preference (M=morning, A=afternoon, E=evening, N=none): ");
                    char c = readCharSafe(sc);
                    Shift primary = Shift.fromChar(c);
                    System.out.print("  Secondary preference? (M/A/E/N) or press Enter for none: ");
                    String secLine = sc.nextLine().trim();
                    Shift secondary = Shift.NONE;
                    if (!secLine.isEmpty()) secondary = Shift.fromChar(secLine.charAt(0));
                    emp.preferences.put(d, new Preference(primary, secondary));
                }
                employees.add(emp);
            }
        }

        // If sample dataset, we may not have filled preferences; ensure each employee has preferences for each day
        fillMissingPreferencesWithRandom(employees);

        // Build schedule
        Schedule schedule = new Schedule();

        // 1st pass: try to assign employees to their primary preferences where possible (respecting MAX capacity and max days)
        for (Day d : Day.week()) {
            // For each shift, gather volunteers
            for (Shift s : new Shift[]{Shift.MORNING, Shift.AFTERNOON, Shift.EVENING}) {
                // gather employees who prefer this shift on this day and are available
                List<Employee> volunteers = employees.stream()
                        .filter(e -> !e.isAssignedOn(d)
                                && e.canWorkMoreDays()
                                && e.preferences.get(d) != null
                                && e.preferences.get(d).primary == s)
                        .collect(Collectors.toList());
                // assign up to MAX_PER_SHIFT from volunteers
                for (Employee e : volunteers) {
                    if (schedule.count(d,s) >= MAX_PER_SHIFT) {
                        // shift is full -> this creates a "conflict" for remaining volunteers
                        break;
                    }
                    schedule.add(d, s, e);
                }
            }
            // After primary assignments, ensure minimum per shift is met by filling from:
            //  - volunteers who had secondary preference for the shift
            //  - employees who are available and under 5 days
            for (Shift s : new Shift[]{Shift.MORNING, Shift.AFTERNOON, Shift.EVENING}) {
                if (schedule.count(d,s) < MIN_PER_SHIFT) {
                    // first pick secondary-preferring employees
                    List<Employee> secondaries = employees.stream()
                            .filter(e -> !e.isAssignedOn(d) && e.canWorkMoreDays()
                                    && e.preferences.get(d) != null
                                    && e.preferences.get(d).secondary == s)
                            .collect(Collectors.toList());
                    Collections.shuffle(secondaries, rand);
                    for (Employee e : secondaries) {
                        if (schedule.count(d,s) >= MIN_PER_SHIFT) break;
                        if (schedule.count(d,s) >= MAX_PER_SHIFT) break;
                        schedule.add(d,s,e);
                    }
                }
                // If still lacking, randomly pick any eligible employee who is not assigned that day and can work more days.
                if (schedule.count(d,s) < MIN_PER_SHIFT) {
                    List<Employee> pool = employees.stream()
                            .filter(e -> !e.isAssignedOn(d) && e.canWorkMoreDays())
                            .collect(Collectors.toList());
                    Collections.shuffle(pool, rand);
                    for (Employee e : pool) {
                        if (schedule.count(d,s) >= MIN_PER_SHIFT) break;
                        if (schedule.count(d,s) >= MAX_PER_SHIFT) break;
                        schedule.add(d,s,e);
                    }
                }
            }
        }

        // SECOND PASS: handle conflicts — employees who preferred a shift but couldn't be placed because shift became full earlier
        // For each employee and each day, if they are unassigned but had a primary preference (not NONE), try to place them:
        for (Day d : Day.week()) {
            for (Employee e : employees) {
                Preference pref = e.preferences.get(d);
                if (pref == null) continue;
                if (!e.isAssignedOn(d) && e.canWorkMoreDays() && pref.primary != Shift.NONE) {
                    // Try secondary on same day
                    boolean placed = false;
                    if (pref.secondary != Shift.NONE) {
                        Shift s = pref.secondary;
                        if (schedule.count(d,s) < MAX_PER_SHIFT) {
                            schedule.add(d,s,e);
                            placed = true;
                        }
                    }
                    // Try other shifts same day
                    if (!placed) {
                        for (Shift s : new Shift[]{Shift.MORNING, Shift.AFTERNOON, Shift.EVENING}) {
                            if (s == pref.primary) continue; // primary already full
                            if (schedule.count(d,s) < MAX_PER_SHIFT) {
                                schedule.add(d,s,e);
                                placed = true;
                                break;
                            }
                        }
                    }
                    // Try next day(s) primary then secondary
                    if (!placed) {
                        for (int offset=1; offset<7 && !placed; ++offset) {
                            Day next = Day.week()[(d.ordinal() + offset) % 7];
                            Preference npref = e.preferences.get(next);
                            if (npref != null && e.canWorkMoreDays() && !e.isAssignedOn(next)) {
                                // try primary on next day
                                if (npref.primary != Shift.NONE && schedule.count(next, npref.primary) < MAX_PER_SHIFT) {
                                    schedule.add(next, npref.primary, e);
                                    placed = true;
                                    break;
                                }
                                // try secondary
                                if (npref.secondary != Shift.NONE && schedule.count(next, npref.secondary) < MAX_PER_SHIFT) {
                                    schedule.add(next, npref.secondary, e);
                                    placed = true;
                                    break;
                                }
                                // try any shift next day
                                for (Shift s2 : new Shift[]{Shift.MORNING, Shift.AFTERNOON, Shift.EVENING}) {
                                    if (schedule.count(next, s2) < MAX_PER_SHIFT) {
                                        schedule.add(next, s2, e);
                                        placed = true;
                                        break;
                                    }
                                }
                            }
                        }
                    }
                    // If still not placed, they will remain unassigned that day (possible if they've reached 5 days or all shifts full)
                }
            }
        }

        // AFTER SCHEDULING: ensure constraints satisfied; if some shift still below MIN_PER_SHIFT, try to fill by relaxing MAX_PER_SHIFT
        for (Day d : Day.week()) {
            for (Shift s : new Shift[]{Shift.MORNING, Shift.AFTERNOON, Shift.EVENING}) {
                if (schedule.count(d,s) < MIN_PER_SHIFT) {
                    // try to find employees who can work (even if it exceeds MAX_PER_SHIFT)
                    List<Employee> pool = employees.stream()
                            .filter(e -> !e.isAssignedOn(d) && e.canWorkMoreDays())
                            .collect(Collectors.toList());
                    Collections.shuffle(pool, rand);
                    for (Employee e : pool) {
                        if (schedule.count(d,s) >= MIN_PER_SHIFT) break;
                        schedule.add(d, s, e);
                    }
                }
            }
        }

        // Final output
        System.out.println("\nFinal Schedule for the Week:");
        System.out.println("============================");
        System.out.println(schedule.prettyPrint());

        // Print employee summaries (how many days assigned + day-by-day)
        System.out.println("Employee assignment summary:");
        for (Employee e : employees) {
            System.out.println("- " + e.name + " -> Assigned days: " + e.assignedDays);
            for (Day d : Day.week()) {
                Shift s = e.assigned.get(d);
                System.out.println("    " + d.name() + ": " + (s == Shift.NONE ? "OFF" : s.name()));
            }
            System.out.println();
        }

        sc.close();
    }

    // Helper: safe read int
    static int readIntSafe(Scanner sc) {
        while (true) {
            String line = sc.nextLine().trim();
            try {
                return Integer.parseInt(line);
            } catch (Exception ex) {
                System.out.print("Please enter a valid integer: ");
            }
        }
    }

    static char readCharSafe(Scanner sc) {
        while (true) {
            String line = sc.nextLine().trim();
            if (line.isEmpty()) return 'N';
            return line.charAt(0);
        }
    }

    // Sample dataset (creates employees and reasonable preferences)
    static List<Employee> sampleEmployees() {
        List<Employee> list = new ArrayList<>();
        Employee a = new Employee("Alice");
        Employee b = new Employee("Bob");
        Employee c = new Employee("Carol");
        Employee d = new Employee("David");
        Employee e = new Employee("Eve");
        Employee f = new Employee("Frank");
        Employee g = new Employee("Grace");

        Collections.addAll(list, a,b,c,d,e,f,g);

        // Example: Alice prefers mornings all week
        for (Day day : Day.week()) a.preferences.put(day, new Preference(Shift.MORNING, Shift.AFTERNOON));
        // Bob prefers afternoons, but off Sundays
        for (Day day : Day.week()) {
            if (day == Day.SUNDAY) b.preferences.put(day, new Preference(Shift.NONE, Shift.NONE));
            else b.preferences.put(day,new Preference(Shift.AFTERNOON, Shift.EVENING));
        }
        // Carol prefers evening mon-fri, morning weekends
        for (Day day : Day.week()) {
            if (day.ordinal() <= Day.FRIDAY.ordinal()-1) // Monday-Friday
                c.preferences.put(day,new Preference(Shift.EVENING, Shift.AFTERNOON));
            else c.preferences.put(day,new Preference(Shift.MORNING, Shift.NONE));
        }
        // David mixed
        a: for (Day day : Day.week()) d.preferences.put(day, new Preference(randomShift(), Shift.NONE));
        // Eve prefers mornings mostly
        for (Day day : Day.week()) e.preferences.put(day, new Preference(Shift.MORNING, Shift.EVENING));
        // Frank prefers afternoon/ evening
        for (Day day : Day.week()) f.preferences.put(day, new Preference(Shift.AFTERNOON, Shift.MORNING));
        // Grace partially unavailable
        for (Day day : Day.week()) {
            if (day == Day.SATURDAY || day == Day.SUNDAY) gracePut(day,g,Shift.NONE,Shift.NONE);
            else gracePut(day,g,Shift.EVENING,Shift.AFTERNOON);
        }

        return list;
    }

    static void gracePut(Day d, Employee g, Shift p, Shift s) {
        g.preferences.put(d, new Preference(p,s));
    }

    static Shift randomShift() {
        int r = rand.nextInt(3);
        return r==0?Shift.MORNING: r==1?Shift.AFTERNOON:Shift.EVENING;
    }

    static void fillMissingPreferencesWithRandom(List<Employee> employees) {
        for (Employee e : employees) {
            for (Day d : Day.week()) {
                if (!e.preferences.containsKey(d)) {
                    // assign a random primary and secondary NONE
                    e.preferences.put(d, new Preference(randomShift(), Shift.NONE));
                }
            }
        }
    }
}
